File: generated_code/GPT4/Python/024_SplayTree.py
class SplayTree:
    class Node:
        __slots__ = ('key', 'left', 'right', 'parent')
        def __init__(self, key, parent=None):
            self.key = key
            self.left = None
            self.right = None
            self.parent = parent

    def __init__(self):
        self.root = None

    def _rotate_left(self, x):
        y = x.right
        if y is None:
            return
        x.right = y.left
        if y.left:
            y.left.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        else:
            if x == x.parent.left:
                x.parent.left = y
            else:
                x.parent.right = y
        y.left = x
        x.parent = y

    def _rotate_right(self, x):
        y = x.left
        if y is None:
            return
        x.left = y.right
        if y.right:
            y.right.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        else:
            if x == x.parent.left:
                x.parent.left = y
            else:
                x.parent.right = y
        y.right = x
        x.parent = y

    def _splay(self, x):
        if x is None:
            return
        while x.parent:
            p = x.parent
            gp = p.parent
            if gp is None:
                # Zig
                if x == p.left:
                    self._rotate_right(p)
                else:
                    self._rotate_left(p)
            else:
                if x == p.left and p == gp.left:
                    # Zig-zig
                    self._rotate_right(gp)
                    self._rotate_right(p)
                elif x == p.right and p == gp.right:
                    # Zig-zig
                    self._rotate_left(gp)
                    self._rotate_left(p)
                elif x == p.right and p == gp.left:
                    # Zig-zag
                    self._rotate_left(p)
                    self._rotate_right(gp)
                elif x == p.left and p == gp.right:
                    # Zig-zag
                    self._rotate_right(p)
                    self._rotate_left(gp)

    def search(self, key):
        node = self.root
        last = None
        while node:
            last = node
            if key == node.key:
                self._splay(node)
                return True
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        if last:
            self._splay(last)
        return False

    def insert(self, key):
        if self.root is None:
            self.root = self.Node(key)
            return
        node = self.root
        parent = None
        while node:
            parent = node
            if key == node.key:
                self._splay(node)
                return  # Already present, do nothing
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        new_node = self.Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        self._splay(new_node)

    def _subtree_minimum(self, node):
        while node.left:
            node = node.left
        return node

    def delete(self, key):
        if self.root is None:
            return
        # Splay key to root (or last accessed node if not found)
        self.search(key)
        if self.root and self.root.key == key:
            if self.root.left is None:
                self.root = self.root.right
                if self.root:
                    self.root.parent = None
            elif self.root.right is None:
                self.root = self.root.left
                if self.root:
                    self.root.parent = None
            else:
                # Find minimum in right subtree
                min_right = self._subtree_minimum(self.root.right)
                if min_right.parent != self.root:
                    self._splay(min_right)
                    min_right.left = self.root.left
                    if min_right.left:
                        min_right.left.parent = min_right
                else:
                    min_right.left = self.root.left
                    if min_right.left:
                        min_right.left.parent = min_right
                self.root = min_right
                self.root.parent = None

    # Optional: for testing purposes
    def __contains__(self, key):
        return self.search(key)

    def __str__(self):
        def _recurse(node):
            if node is None:
                return []
            return _recurse(node.left) + [node.key] + _recurse(node.right)
        return str(_recurse(self.root))

File: generated_code/GPT4/Java/024_SplayTree.java
public class SplayTree {
    private class Node {
        int key;
        Node left, right;

        Node(int key) {
            this.key = key;
        }
    }

    private Node root;

    // Splay operation: moves the node with the given key to the root (or the last accessed node)
    private Node splay(Node root, int key) {
        if (root == null || root.key == key) {
            return root;
        }

        // Key lies in left subtree
        if (key < root.key) {
            if (root.left == null) return root;

            // Zig-Zig (Left Left)
            if (key < root.left.key) {
                // First recursively bring the key as root of left-left
                root.left.left = splay(root.left.left, key);
                // First rotation for root
                root = rightRotate(root);
            }
            // Zig-Zag (Left Right)
            else if (key > root.left.key) {
                // Bring the key as root of left-right
                root.left.right = splay(root.left.right, key);
                // If left-right child exists, do first rotation
                if (root.left.right != null) {
                    root.left = leftRotate(root.left);
                }
            }

            // Second rotation for root
            return (root.left == null) ? root : rightRotate(root);
        }
        // Key lies in right subtree
        else {
            if (root.right == null) return root;

            // Zag-Zig (Right Left)
            if (key < root.right.key) {
                // Bring the key as root of right-left
                root.right.left = splay(root.right.left, key);
                // If right-left child exists, do first rotation
                if (root.right.left != null) {
                    root.right = rightRotate(root.right);
                }
            }
            // Zag-Zag (Right Right)
            else if (key > root.right.key) {
                // Bring the key as root of right-right
                root.right.right = splay(root.right.right, key);
                root = leftRotate(root);
            }

            // Second rotation for root
            return (root.right == null) ? root : leftRotate(root);
        }
    }

    // Right rotate
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    // Left rotate
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    // Public insert method
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }
        root = splay(root, key);
        if (root.key == key) {
            // Key already exists, do nothing (set does not allow duplicates)
            return;
        }
        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else {
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    // Public delete method
    public void delete(int key) {
        if (root == null) return;
        root = splay(root, key);
        if (root.key != key) return; // Key not found

        // Now root.key == key
        if (root.left == null) {
            root = root.right;
        } else {
            Node temp = root.right;
            root = root.left;
            // Splay the maximum node in the left subtree to the root
            root = splay(root, key);
            // Attach the right subtree
            root.right = temp;
        }
    }

    // Public search method. Returns Integer key if found, else null
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    // Optional: For testing. In-order traversal
    public void inorder() {
        inorder(root);
        System.out.println();
    }

    private void inorder(Node node) {
        if (node == null) return;
        inorder(node.left);
        System.out.print(node.key + " ");
        inorder(node.right);
    }
}